hardware/VirtualThermostat.cpp

```diff
@@ -0,0 +1,682 @@
+#include "stdafx.h"
+#include "../main/Helper.h"
+#include "../main/Logger.h"
+#include "../main/SQLHelper.h"
+#include "../main/RFXtrx.h"
+
+#include "VirtualThermostat.h"
+
+#include "../main/localtime_r.h"
+#include "../main/mainworker.h"
+#include "../notifications/NotificationHelper.h"
+#include <stdarg.h>     
+
+std::string AVALAIBLE_MODE = "Eco,Conf,Frost,Off";
+
+std::string GetDeviceValue(const char* FieldName, const char* Idx)
+{
+	auto result = m_sql.safe_query("SELECT %s from DeviceStatus WHERE (ID == %s )", FieldName, Idx);
+	if (!result.empty())
+		return  result[0][0];
+	else
+		return  "";
+}
+
+void VirtualThermostat::CircularBuffer::Clear()
+{
+	for (int i = 0; i < m_Size; i++)m_Value[i] = 0;
+	m_index = 0;
+}
+VirtualThermostat::CircularBuffer::CircularBuffer(int pSize)
+{
+	m_Value = new double[pSize];
+	m_Size = pSize;
+	Clear();
+}
+VirtualThermostat::CircularBuffer::~CircularBuffer()
+{
+	delete[] m_Value;
+}
+int VirtualThermostat::CircularBuffer::GetNext()
+{
+	if (m_index >= (m_Size - 1))
+		return 0;
+	else
+		return m_index + 1;
+}
+double VirtualThermostat::CircularBuffer::Put(double val)
+{
+	double lastv = m_Value[m_index];
+	m_Value[m_index] = val;
+	m_index = GetNext();
+	return lastv;
+}
+
+double VirtualThermostat::CircularBuffer::GetSum()
+{
+	double  Sum = 0;
+	for (int i = 0; i < m_Size; i++)
+	{
+		Sum += m_Value[i];
+	}
+	return Sum;
+}
+
+//option management
+void getOption(TOptionMap& Option, const char* optionName, int& value)
+{
+	value = std::stoi(Option[optionName].c_str());
```

Comment:
function std::stoi or std::stol can rise an std::exception. Is that ok?  and std::stoi can work with std::string